pkg/hive/feature_lifecycle.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package hive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+
+	"github.com/cilium/hive/cell"
+	"golang.org/x/exp/maps"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type FeatureLifecycleInterface interface {
+	Append(Feature, cell.Hook) error
+	Start(Feature, context.Context, *slog.Logger) error
+	Stop(Feature, context.Context, *slog.Logger) error
+
+	IsRunning(Feature) bool
+	List() []Feature
+}
+
+type Feature string
+type FeatureLifecycle struct {
+	mu     lock.Mutex
+	hooks  map[Feature][]cell.Hook
+	status map[Feature]bool
+}
+
+func NewFeatureLifecycle() *FeatureLifecycle {
+	return &FeatureLifecycle{
+		hooks:  make(map[Feature][]cell.Hook),
+		status: make(map[Feature]bool),
+	}
+}
+
+// Append adds a hook to the feature hooks, marking the feature as not running.
+// It returns an error if the feature is already running.
+func (fl *FeatureLifecycle) Append(f Feature, h cell.Hook) error {
+	fl.mu.Lock()
+	defer fl.mu.Unlock()
+
+	if status, ok := fl.status[f]; ok && status {
+		return fmt.Errorf("cannot add hooks to a running feature: %s", f)
+	}
+
+	fl.hooks[f] = append(fl.hooks[f], h)
+	fl.status[f] = false
+
+	return nil
+}
+
+// Start attempts to start a feature by executing its associated hooks.
+// It returns an error if the feature is already running
+// or if any hook fails to start.
+func (fl *FeatureLifecycle) Start(f Feature, c context.Context, l *slog.Logger) error {
+	fl.mu.Lock()
+	defer fl.mu.Unlock()
+
+	if status, ok := fl.status[f]; ok && status {
+		return fmt.Errorf("feature %s is already running", f)
+	}
+
+	l.Debug("Starting hooks for", "feature", f)
+	for _, hook := range fl.hooks[f] {
+		if err := hook.Start(c); err != nil {
+			l.Error("Start hook failed", "error", err)
+			return fmt.Errorf("starting hook for feature %s: %w", f, err)
+		}
+	}
+
+	fl.status[f] = true
+
+	return nil
+}
+
+// Stop attempts to stop a feature by stopping its associated hooks.
+// It returns an error if the feature is already stopped.
+// If any hook encounters an error during stopping it aggregates into return error
+func (fl *FeatureLifecycle) Stop(f Feature, c context.Context, l *slog.Logger) error {
```

Comment:
How are you expecting this to interact with the normal lifecycle? The code as it stands seems to imply that `FeatureLifecycle` is provided to the hive, that cells will take it as parameter instead of `cell.Lifecycle` and attach their start hook to a particular feature.

Whenever a trigger happens like starting with a particular flag enabled or some sort of dynamic config update, you would call `.Start`.

But the `FeatureLifecycle` itself doesn't seem aware of when the main hive is shutting down. That could cause improper shutdown. So are users if this feature expected to also append the same stop hook to the `cell.Lifecycle` or are you planning on adding logic that call stop for all running features upon shutdown?

Reply:
A new LifecycleManager is work in progress and should manage this FeatureLifecycle, the manager should also consider the dependencies between features and have the logic to no allow the user to stop features that are being used by other features as a hard dependency.

I should've added this detail in the PR description from beginning. I'm sorry for the confusion.